Infinite scroll for People Management

oguzkocer opened the pull request:
Relevant #3901. This PR adds infinite scroll to people list. As per our discussion on Slack, we remove the users list for a site when you visit the people page, except for the first 20 users so we don't show an empty list. The idea behind this implementation is to make sure the removed users don't show up in that list.

When a fresh list of users is received from remote, meaning the request `offset` parameter is `0`, we remove ALL users for that site and fill it with the users from the response. From then on, when you scroll down to the end of list, a new page will be requested and we'll just save those users on top of the ones we already have. I've implemented the same progress bar we use in the reader posts page to show that we're loading more users. I just centered it because it was blocking the gravatar for the users (it's on the left for reader).

I moved `PeopleAdapter` into `PeopleListFragment` to handle end of list condition a bit easier. Initially I tried to add a listener to the adapter which was the fragment and that'd trigger the activity to make the request, but then I thought it was too long of a chain. Moving the adapter inside of the fragment made things simpler.

Question:

* In `deletePeopleForLocalBlogIdExceptForFirstPage` method of `PeopleTable` I was unable to add the table's name `people` as a query parameter. The query to delete `n` users from the end of a table turned out to be a bit complicated. I'd appreciate any suggestions on improving that, as well as finding a way to add the `PEOPLE_TABLE` as a parameter.

Known issue:

* This is unrelated to this PR but I've recently found out that `public View getView(int position, View convertView, ViewGroup parent)` is called 3 times for each row in `PeopleAdapter`. I was unable to find the source of the issue so far and will be handling it separately.

To test: Navigate to a blog where there is more than 20 users, go into People page and scroll to the end of list. You should see a loading progress bar and once that's gone, you should be able to scroll to see more users.

/cc @astralbodies, @hypest, @jleandroperez 

Needs review: @nbradbury

nbradbury commented on WordPress/src/main/java/org/wordpress/android/datasets/PeopleTable.java:
```diff
@@ -80,11 +83,28 @@ public static void savePeople(List<Person> peopleList, int localTableBlogId) {
+        if (size > fetchLimit) {
+            int deleteCount = size - fetchLimit;
+            String[] args = new String[]{Integer.toString(deleteCount), Integer.toString(localTableBlogId)};
+            getWritableDb().delete(PEOPLE_TABLE, "person_id " +
```
Rather than hard-code "people" here, how about `"person_id IN (SELECT person_id FROM " + PEOPLE_TABLE + "WHERE..."`

oguzkocer replied:
I've added a question to the PR description just for this 👊, because I was trying to use `?` and pass the `PEOPLE_TABLE` as an argument. Working on that sql statement probably fried my brain a bit. 😢

nbradbury commented on WordPress/src/main/java/org/wordpress/android/datasets/PeopleTable.java:
```diff
@@ -80,11 +83,28 @@ public static void savePeople(List<Person> peopleList, int localTableBlogId) {
+            int deleteCount = size - fetchLimit;
+            String[] args = new String[]{Integer.toString(deleteCount), Integer.toString(localTableBlogId)};
+            getWritableDb().delete(PEOPLE_TABLE, "person_id " +
+                    "IN (SELECT person_id FROM people ORDER BY display_name DESC LIMIT ?) AND local_blog_id=?", args);
```
The `IN` statement should have a `WHERE` clause using `local_blog_id` (right now it's selecting all people regardless of blog).

nbradbury commented:
> This is unrelated to this PR but I've recently found out that public View getView(int position, View convertView, ViewGroup parent) is called 3 times for each row in PeopleAdapter. I was unable to find the source of the issue so far and will be handling it separately.

I remember running into this same problem in the past. Replacing the `ListView` with a `RecyclerView` should fix it.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleListFragment.java:
```diff
@@ -1,24 +1,34 @@
 import java.util.List;
 
 public class PeopleListFragment extends ListFragment implements OnItemClickListener {
     private static String ARG_LOCAL_TABLE_BLOG_ID = "local_table_blog_id";
```
Just noticed this should be `final`.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleListFragment.java:
```diff
@@ -72,16 +93,120 @@ public void refreshPeopleList() {
+            }
+
+            // end of list is reached
+            if (mOnFetchMorePeopleListener != null && position == getCount() - 1) {
```
This (or the related code in the activity) should prevent fetching users when a fetch is already in progress. Otherwise the user could scroll up and down, causing multiple fetches for the same offset.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleManagementActivity.java:
```diff
@@ -256,4 +301,14 @@ private Person getCurrentPerson() {
+        if (mPeopleEndOfListReached) {
+            return;
+        }
+        Blog blog = WordPress.getCurrentBlog();
```
Since this only needs the blog's ID, it could use `WordPress.getCurrentLocalTableBlogId()`

oguzkocer replied:
I really should just use Github instead of going through the emails.. I came here to say the same thing just to see that you've already edited it 👍

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/utils/PeopleUtils.java:
```diff
@@ -46,8 +49,13 @@ public void onErrorResponse(VolleyError volleyError) {
+        Map<String, String> params = new HashMap<>();
+        params.put("number", Integer.toString(PeopleUtils.FETCH_USERS_LIMIT));
+        params.put("offset", Integer.toString(offset));
+        params.put("order_by", "display_name");
```
Although [the docs](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/users/) do say `display_name` is a valid `order_by` parameter, I don't see `display_name` in the response. Are we sure this is correct?

oguzkocer replied:
Yes! This is very weird indeed. In the response we have `name` as the display name, however if you use the [console](https://developer.wordpress.com/docs/api/console/) to `order_by` the `name` it doesn't order it the way we want. You actually need to use `display_name` even though it's not in the response at all and what's more is that the default order is descending. 😕 

We should actually ping @ebinnion in case this is something we want to improve on the API side.

ebinnion replied:
I believe this is because WordPress uses `display_name` for the query, but we were attempting to change it to `name` for the API. I would suggest creating a patch to map `name` to `display_name`.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleManagementActivity.java:
```diff
@@ -85,16 +110,36 @@ public boolean onOptionsItemSelected(final MenuItem item) {
+        FragmentManager fragmentManager = getFragmentManager();
+        final PeopleListFragment peopleListFragment = (PeopleListFragment) fragmentManager
+                .findFragmentByTag(KEY_PEOPLE_LIST_FRAGMENT);
+        if (peopleListFragment != null && offset > 0) {
```
There's no loading indicator the first time the people list is shown, so user is left looking at an empty screen until the request completes. Maybe there should be a progress bar when the list is empty?

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleListFragment.java:
```diff
@@ -72,16 +93,120 @@ public void refreshPeopleList() {
+            if (person == null) {
+                return -1;
+            }
+            return person.getPersonID();
```
Since each person's ID is different, the list can be optimized by returning `true` for `hasStableIds()`, ex:
```
@Override
public boolean hasStableIds() {
    return true;
}
```
Explanation [here](http://stackoverflow.com/a/29446944/1673548).

nbradbury replied:
Hmm, I'm wondering simply adding that `hasStableIds()` change would make a difference with that problem you're seeing with multiple `getView()` calls?

nbradbury commented on WordPress/src/main/java/org/wordpress/android/datasets/PeopleTable.java:
```diff
@@ -99,9 +99,14 @@ public static void deletePeopleForLocalBlogIdExceptForFirstPage(int localTableBl
+                    Integer.toString(localTableBlogId),
+                    Integer.toString(deleteCount)
+            };
+            getWritableDb().delete(PEOPLE_TABLE, "local_blog_id=? AND person_id " +
```
For future reference, you can use numbered arguments (like `?1`) in queries so you don't need to pass the same argument twice. Example:

```
String[] args = new String[] {
        Integer.toString(localTableBlogId),
        Integer.toString(deleteCount)
};
getWritableDb().delete(PEOPLE_TABLE, "local_blog_id=?1 AND person_id " +
        "IN (SELECT person_id FROM " + PEOPLE_TABLE + " WHERE local_blog_id=?1 " +
        "ORDER BY display_name DESC LIMIT ?2)", args);
```

oguzkocer replied:
TIL, thanks!

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleListFragment.java:
```diff
@@ -189,6 +179,15 @@ public void onBindViewHolder(PeopleAdapter.PeopleViewHolder holder, int position
+            holder.itemView.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    if (mOnPersonSelectedListener != null) {
```
We handle click listeners in recyclers this same way in much of our code (and I'm responsible for some of them), but it turns it's not always safe - there are cases where relying on a `final` variable this way can result in the click listener using the wrong item.

A safer way is for the listener to use `getAdapterPosition()` to locate the position of the clicked view, and then get the model (a `Person` in this case) from that position. An example of this can be seen in our [CommentAdatper](https://github.com/wordpress-mobile/WordPress-Android/blob/feature/people-infinite-scroll/WordPress/src/main/java/org/wordpress/android/ui/comments/CommentAdapter.java#L102-L118).

nbradbury commented:
Nice work on the switch to a `RecyclerView` - but unfortunately the change seems to have broken infinite scroll. I can no longer scroll to get more than the initial 20 users.

nbradbury commented:
Ignore my previous comment about infinite scroll not working - that was apparently caused by some code changes I made while testing. Sorry for the false alarm!

oguzkocer commented:
@nbradbury I've addressed all your feedback. fdec9bef803ad5fd655ba08f56fb2ef0810e063b is the best way I could find that ensures the progress bar is visible at correct times. The PR should be ready for another pass. Thanks!

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleListFragment.java:
```diff
@@ -157,6 +157,15 @@ public int getItemCount() {
         }
 
         @Override
+        public long getItemId(int position) {
```
In addition to this, the adapter's constructor should have `setHasStableIds(true);`

oguzkocer replied:
I swear I had that! It didn't even work at first when I added it because I removed the `getItemId` while I was changing to `RecyclerView`. I guess I forgot to put it back in, so sorry, fixing now!

oguzkocer commented:
@nbradbury I've improved the offline mode in 821625e8c856ce9c169f69fba494f7c2c5c38536. We don't remove the cached users if there is no connectivity anymore (as you suggested) and we don't try to make a call to the server for any action. I've checked our codebase and generally we silently fail if there is no connection, but I think it'd be better if we tell the user about it. For now, I've used the same errors we use if the request actually fails, however I feel like it'd be better if we had a general error string to use in these situations, maybe something like "No network connection". What do you think?

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleManagementActivity.java:
```diff
@@ -177,6 +181,11 @@ public void onPersonSelected(Person person) {
     @Override
     public void onRoleChanged(long personID, int localTableBlogId, String newRole) {
+        if(!NetworkUtils.isNetworkAvailable(this)) {
+            ToastUtils.showToast(this, R.string.error_update_role, ToastUtils.Duration.LONG);
```
Since the error is due to no network, it'd be nicer to tell the user that's what the problem is. How about `R.string.no_network_message` instead?

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleManagementActivity.java:
```diff
@@ -232,6 +241,11 @@ public void onClick(DialogInterface dialog, int which) {
 
     private void removeSelectedPerson() {
+        if(!NetworkUtils.isNetworkAvailable(this)) {
+            ToastUtils.showToast(this, R.string.error_remove_user, ToastUtils.Duration.LONG);
```
Same comment as above - `no_network_message` would be friendlier.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleManagementActivity.java:
```diff
@@ -55,7 +56,10 @@ public void onCreate(Bundle savedInstanceState) {
         if (savedInstanceState == null) {
-            PeopleTable.deletePeopleForLocalBlogIdExceptForFirstPage(blog.getLocalTableBlogId());
+            // only delete cached people if there is a connection
+            if (NetworkUtils.isNetworkAvailable(this)) {
```
I think it would make sense to use `NetworkUtils.checkConnection()` here instead, since that shows a toast letting the user know there's no connection. Without this, the user may be stuck looking at an empty list with no indication of why it's empty (and the message is useful even if cached users are shown).

oguzkocer replied:
So, this is kind of what I was asking in my comment [here](https://github.com/wordpress-mobile/WordPress-Android/pull/4120#issuecomment-221888645). However, I don't agree with showing any messages here. We'll be trying to fetch a fresh list of users and if network connection is not available we already show a message there.

Since this is mostly a hack to make sure removed users in remote are removed in our cache as well, I wouldn't do anything UI related here and will defer to the fetching method to inform the user.

nbradbury replied:
> So, this is kind of what I was asking in my comment here.

Bah, sorry about that. I pulled the branch without realizing you'd commented.

jleandroperez commented on WordPress/src/main/java/org/wordpress/android/ui/people/utils/PeopleUtils.java:
```diff
@@ -46,8 +49,13 @@ public void onErrorResponse(VolleyError volleyError) {
         };
 
+        Map<String, String> params = new HashMap<>();
+        params.put("number", Integer.toString(PeopleUtils.FETCH_USERS_LIMIT));
```
Guys, heads up. Not sure if we're hitting the exact same endpoint, or there's a difference. But on my end, the `number` parameter appears to be disregarded.

Instead, i'm using `page`. (Checked with dotcom source!).

jleandroperez replied:
Okay. Update. Apparently *Users* deal with `offset`, while *Followers* deal with `page`

oguzkocer replied:
Thanks for the heads up @jleandroperez. I am currently working on the follower section and I'll make sure we use the correct parameters for this. We currently use the default values, so probably that's why I didn't notice the issue before and just used what the documentation said here: https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/users/
